File: 2022/python/src/solutions/day_07.py
from contextlib import suppress
from functools import cache

from src.lib.solution import SolutionABC


class Solution(SolutionABC):
    @cache
    def _common(self) -> list[int]:
        tmp = []
        directories = []
        for line in self._data.splitlines():
            match line.split():
                case ["$", "cd", ".."]:
                    directories.append(tmp.pop()),
                case ["$", "cd", _]:
                    tmp.append(0),
                case [size, _] if size.isdigit():
                    tmp = [num + int(size) for num in tmp]

        directories += tmp

        return directories

    def _part_a(self) -> str:
        result = sum([num for num in self._common() if num < 100_000])

        return str(result)  # 2031851

    def _part_b(self) -> str:
        root = max(self._common())
        required = root + 30_000_000 - 70_000_000
        result = min([num for num in self._common() if num >= required])

        return str(result)  # 2568781

File: 2022/rust/src/solution/day_07.rs
use crate::common::io::read_day;
use crate::common::solution::Solution;

pub struct Day {
    directories: Vec<u32>,
}

impl Day {
    pub fn new() -> Self {
        let mut tmp_dir: Vec<u32> = Vec::new();
        let mut directories: Vec<u32> = Vec::new();

        for line in read_day(7).unwrap().lines() {
            match line.split(' ').collect::<Vec<&str>>().as_slice() {
                ["$", "cd", ".."] => directories.push(tmp_dir.pop().unwrap()),
                ["$", "cd", _] => tmp_dir.push(0),
                [size, _] => {
                    // Getting rid of "dir ..." and "$ ls" here
                    if let Ok(num) = size.parse::<u32>() {
                        tmp_dir.iter_mut().for_each(|n| *n += num)
                    }
                }
                [..] => unreachable!(),
            }
        }
        directories.extend(tmp_dir);

        Self { directories }
    }
}

impl Solution for Day {
    fn name(&self) -> &'static str { "Day 7: No Space Left On Device" }

    fn part_a(&self) -> Option<String> {
        let result = self.directories
                .iter()
                .filter(|&&size| size < 100_000)
                .sum::<u32>();

        Some(result.to_string()) // 2031851

    }

    fn part_b(&self) -> Option<String> {
        let root = *self.directories.iter().max().unwrap();  // Biggest directory
        let required = root + 30_000_000 - 70_000_000;  // Required size for dir
        let result = self.directories
            .iter()
            .filter(|&&dir| dir >= required)
            .min()
            .unwrap();

        Some(result.to_string())  // 2568781
    }
}

